cmd/snap-seccomp/main.go

```diff
@@ -0,0 +1,522 @@
+	}
+
+	return strconv.ParseUint(token, 10, 64)
+}
+
+func parseLine(line string, secFilter *seccomp.ScmpFilter) error {
+	line = strings.TrimSpace(line)
+
+	// ignore comments and empty lines
+	if strings.HasPrefix(line, "#") || line == "" {
+		return nil
+	}
+
+	// regular line
+	tokens := strings.Fields(line)
+
+	// fish out syscall
+	secSyscall, err := seccomp.GetSyscallFromName(tokens[0])
+	if err != nil {
+		// FIXME: add structed error to libseccomp-golang
+		if err.Error() == "could not resolve name to syscall" {
+			return nil
+		}
+		return fmt.Errorf("cannot resolve name: %s", err)
+	}
+
+	var conds []seccomp.ScmpCondition
+	if err != nil {
+		return fmt.Errorf("cannot create new filter: %s", err)
+	}
+
+	for pos, arg := range tokens[1:] {
+		var cmpOp seccomp.ScmpCompareOp
+		var value uint64
+		var err error
+
+		if pos >= ScArgsMaxlength {
+			return fmt.Errorf("too many tokens (%d) in line %q", pos, line)
+		}
+
+		if arg == "-" {
+			continue
+		}
+
+		if strings.HasPrefix(arg, ">=") {
+			cmpOp = seccomp.CompareGreaterEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "<=") {
+			cmpOp = seccomp.CompareLessOrEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "!") {
+			cmpOp = seccomp.CompareNotEqual
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "<") {
+			cmpOp = seccomp.CompareLess
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, ">") {
+			cmpOp = seccomp.CompareGreater
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "|") {
+			cmpOp = seccomp.CompareMaskedEqual
+			value, err = readNumber(arg[1:])
+		} else {
+			cmpOp = seccomp.CompareEqual
+			value, err = readNumber(arg)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse token %q (line %q)", arg, line)
+		}
+
+		var scmpCond seccomp.ScmpCondition
+		if cmpOp == seccomp.CompareMaskedEqual {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value, value)
+		} else {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse line %q: %s", line, err)
+		}
+		conds = append(conds, scmpCond)
+	}
+
+	// FIXME: why do we need this fallback?
+	if err = secFilter.AddRuleConditionalExact(secSyscall, seccomp.ActAllow, conds); err != nil {
+		err = secFilter.AddRuleConditional(secSyscall, seccomp.ActAllow, conds)
+	}
+
+	return err
+}
+
+// For architectures that support a compat architecture, when the
+// kernel and userspace match, add the compat arch, otherwise add
+// the kernel arch to support the kernel's arch (eg, 64bit kernels with
+// 32bit userspace).
+func addSecondaryArches(secFilter *seccomp.ScmpFilter) error {
+	switch arch.UbuntuArchitecture() {
+	case "amd64":
+		return secFilter.AddArch(seccomp.ArchX86)
+	case "arm64":
+		return secFilter.AddArch(seccomp.ArchARM)
+	case "ppc64el":
+		return secFilter.AddArch(seccomp.ArchPPC)
+	}
+	return nil
```

Comment:
64 bit kernels support filtering both their native syscalls and compat syscalls. Consider the following seccomp denial:
```
kernel: [1288522.491300] audit: type=1326 audit(1497441564.893:222342): auid=1000 uid=0 gid=0 ses=3 pid=30492 comm="curl" exe="/usr/bin/curl" sig=31 arch=c000003e syscall=41 compat=0 ip=0x7f2f1a8ec387 code=0x0
```

The `compat=0` indicates that the syscall is using the kernel's native syscall. If you run an 32 bit x86 binary on amd64, this would be `compat=1`. To support syscall filtering for compat architectures, we need to load the compat architecture for that kernel. libseccomp will by default only load the architecture that matches its architecture (eg, if libseccomp is 32 bit, it loads that, if 64, it loads that). 

Therefore to fully support compat syscalls, we need to handle:
- kernel and userspace are both 32 bit (no compat)
- kernel and userspace are both 64 bit (need to load 32 bit arch for 32 bit compat)
- kernel is 64 bit and userspace is 32 bit (need to load 64 bit arch for native syscalls)

32 bit kernel and 64 bit userspace is of course not supported. The C implementation handles all 3 of the above cases, this PR currently only handles the first two. I'm told the 3rd case is interesting in embedded where it is not uncommon to run a 64 bit kernel but otherwise use a 32 bit userspace (it isn't a typical general Linux distro combination, but snapd is going everywhere, so we need to support it).